apps/opt.c

```diff
@@ -676,25 +680,52 @@ int opt_next(void)
             /* Just a string. */
             break;
         case '/':
-            if (app_isdir(arg) >= 0)
+            if (app_isdir(arg) > 0)
                 break;
             BIO_printf(bio_err, "%s: Not a directory: %s\n", prog, arg);
             return -1;
         case '<':
             /* Input file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) >= 0)
+            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) == 0)
                 break;
             BIO_printf(bio_err,
                        "%s: Cannot open input file %s, %s\n",
                        prog, arg, strerror(errno));
             return -1;
         case '>':
             /* Output file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, W_OK) >= 0 || errno == ENOENT)
+#if !defined(_WIN32) && !defined(__VMS)
+            c = OPENSSL_strdup(arg);
+            if (c == NULL) {
+                BIO_printf(bio_err,
+                           "%s: Memory allocation failure\n", prog);
+                return -1;
+            }
+            oerrno = errno;
+            errno = 0;
+            if (strcmp(arg, "-") == 0
+                || (app_access(app_dirname(c), W_OK) == 0
+                    && app_isdir(arg) <= 0
+                    && (app_access(arg, W_OK) == 0 || errno == ENOENT))) {
+                OPENSSL_free(c);
                 break;
+            }
+            OPENSSL_free(c);
+            if (errno == 0)
+                /* only possible if 'arg' is a directory */
+                estr = "is a directory";
+            else
+                estr = strerror(errno);
+            errno = oerrno;
+#else
+            if (strcmp(arg, "-") == 0 || app_access(arg, W_OK) == 0
+                || errno == ENOENT)
```

Comment:
ENOENT is a POSIX thing. What if we are running on a non POSIX platform?

Reply:
```
: ; git grep ENOENT
CHANGES:     ERR_R_SYS_LIB error code if fopen() fails with ENOENT.
apps/apps.c:    if (rename(serialfile, buf[1]) < 0 && errno != ENOENT
apps/apps.c:    if (rename(dbfile, buf[1]) < 0 && errno != ENOENT
apps/apps.c:    if (rename(buf[4], buf[3]) < 0 && errno != ENOENT
apps/apps.c:        errno = ENOENT;
apps/opt.c:            if (strcmp(arg, "-") == 0 || app_access(arg, W_OK) >= 0 || errno == ENOENT)
apps/rehash.c:                    if (unlink(buf) < 0 && errno != ENOENT) {
apps/rehash.c:                    if (unlink(buf) < 0 && errno != ENOENT) {
crypto/LPdir_vms.c:            errno = ENOENT;
crypto/LPdir_win.c:            errno = ENOENT;
crypto/bio/bss_file.c:        if (errno == ENOENT
crypto/o_fopen.c:            (errno == ENOENT || errno == EBADF)
```

It exists outside of POSIX.

Reply:
Ok